jina/serve/runtimes/head/__init__.py

```diff
@@ -329,10 +332,26 @@ async def _handle_data_request(
             timeout=self.timeout_send,
             retries=self._retries,
         )
+        total_shards = len(worker_send_tasks)
 
-        worker_results = await asyncio.gather(*worker_send_tasks)
+        all_worker_results = await asyncio.gather(*worker_send_tasks)
+        worker_results = list(
+            filter(lambda x: isinstance(x, Tuple), all_worker_results)
+        )
+        exceptions = list(
+            filter(
+                lambda x: isinstance(x, (AioRpcError, InternalNetworkError)),
+                all_worker_results,
+            )
+        )
+        failed_shards = len(exceptions)
+        if failed_shards:
```

Comment:
Just a suggestion, bu personally I would pull this into a separate method, seeing that `_handle_data_request()` is already quite long.

Reply:
I would have also done this automatically but the tests are not working reliably on macOS. This is a huge productivity sink. I basically spent 80% of my time on this pr trying to run tests locally. Anyway good point and I have more work to ensure that Jina runs correctly on macOS.